### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by lauraguerra1**
Description
<br>
- This PR introduces a search bar and search functionality for the main page 
Screen Shot 2023-05-19 at 3 22 01 PM
Screen Shot 2023-05-19 at 3 22 51 PM

Testing Notes
<br>
- New tests are implemented for a function `searchRecipes`  that combines two previous filtering functions 
- As a user, you should be able to search for ingredients and recipe names at the same time
- The search of `wheat cake` should result in any recipes that have cake or wheat in their name or ingredients 
- Searches are not case-sensitive. 
- If the search does not match any results, the user should see an apologetic message

### CODE REVIEW

**Comment by lauraguerra1** [comment]
[[Iteration 1 & 2]Search functionality](https://trello.com/c/Itb4QmVX/12-iteration-1-2search-functionality)

File: src/styles.css [hunk 1]
```diff
@@ -371,9 +407,20 @@ input {
...
 }
+/* }
+
+#ingredientsList::-webkit-scrollbar-track {
+  background-color: #3b4154;
+}
+
+#ingredientsList::-webkit-scrollbar-thumb {
+  background-color: #fbbf83;
+} */
```
Comment (mbenfowler) [inline src/styles.css]: what's going on here?

File: src/recipes.js [hunk 2]
```diff
@@ -35,24 +35,33 @@ const calculateRecipeCost = (recipe, ingredients) => {
...
   return matchedRecipes;
 };
 
+const searchRecipes = (allRecipes, allIngredients, userSearch) => {
+  const searchedItems = userSearch.split(' ')
+  const recipesByName = searchedItems.flatMap(searchWord => filterRecipesByName(allRecipes, searchWord))
+  const recipesByIngredient = searchedItems.flatMap(searchWord => filterRecipesByIngredient(allRecipes, searchWord, allIngredients))
+  const foundRecipes = [...recipesByName, ...recipesByIngredient]
+  return [...new Set(foundRecipes)]
+}
```
Comment (mbenfowler) [inline src/recipes.js]: incredible 👏

File: src/domUpdates.js [hunk 3]
```diff
@@ -1,8 +1,11 @@
 // Imports
 import {recipeData} from './data/recipes'
-import {recipeGrid, clickedRecipe, tagArea, getRecipeCard, allRecipes, ingredientsList} from './scripts'
+import { ingredientsData } from './data/ingredients';
+import {recipeGrid, clickedRecipe, tagArea, getRecipeCard, allRecipes, ingredientsList, searchBar} from './scripts'
+import { searchRecipes } from './recipes';
 
 let currentRecipeCard = require('./scripts');
+let searchedRecipes = require('./scripts')
```
Comment (mbenfowler) [inline src/domUpdates.js]: so I'm playing around with this in my data model card - looks like we may not need these two lines?

File: src/scripts.js [hunk 4]
```diff
@@ -72,13 +77,17 @@ allRecipes.addEventListener("click", (event) => {
...
+searchBtn.addEventListener('click', searchForRecipes)
+
 // Exports
 export {
   recipeGrid,
   tagArea,
   clickedRecipe,
   getRecipeCard,
   currentRecipeCard,
+  searchedRecipes,
```
Comment (mbenfowler) [inline src/scripts.js]: again, we may not need to export these.  We already have access to the functions in this file that will be manipulating our DM.

File: test/recipe-test.js [hunk 5]
```diff
@@ -289,3 +290,39 @@ describe('filterRecipes', () => {
...
+    let expectedRecipesByName = [sampleRecipeData[3], sampleRecipeData[4]];
+    let filteredRecipesByIngredient = searchRecipes(sampleRecipeData, sampleIngredientsData, 'wheat')
+    let expectedRecipesByIngredient = [sampleRecipeData[0],sampleRecipeData[3],sampleRecipeData[4]];
+    
+    assert.deepEqual(filteredRecipesByName, expectedRecipesByName);
+    assert.deepEqual(filteredRecipesByIngredient, expectedRecipesByIngredient);
+  })
+
+  it('should not be case sensitive', () => {
+    let filteredRecipes = searchRecipes(sampleRecipeData, sampleIngredientsData, 'cAkE')
```
Comment (mbenfowler) [inline test/recipe-test.js:311]: gIvE mE cAkE NOWWWW

File: test/recipe-test.js [hunk 6]
```diff
@@ -289,3 +290,39 @@ describe('filterRecipes', () => {
     expect(filteredRecipes).to.deep.equal(expectedRecipes);
   })
 });
+
+describe('search recipes', () => {
```
Comment (mbenfowler) [inline test/recipe-test.js:294]: Love all these tests!!! Maybe we should have one for what happens when a user clicks the search button when no search term is entered in the search box?  The current behavior right now is that nothing happens.

Maybe connectedly, there is also no way to "reset" the search back to the initial state with all of the recipes?

**Review by mbenfowler — CHANGES_REQUESTED** [review]
Made some comments!  Also, I think this is non-blocking for this one, but when only 1 or 2 recipes come back from our search it kind of destroys our nice 3-column approach:
Screenshot 2023-05-19 at 4 03 21 PM

File: test/recipe-test.js (hunk 5 above)
Comment (taranveersingh93) [reply test/recipe-test.js:311]: Lovely edge case!

File: dist/index.html [hunk 7]
```diff
@@ -8,6 +8,12 @@
     <title>Whats Cookin</title>
   </head>
   <body>
+    <nav>
```
Comment (taranveersingh93) [inline dist/index.html:11]: nice semantics

File: src/domUpdates.js [hunk 8]
```diff
@@ -58,8 +61,8 @@ const createGridHTML = allColumns => {
   return htmlCode;
 }
 
-const renderGrid = () => {
-  const gridData = makeRecipeColumnData(recipeData)
+const renderGrid = (data) => {
```
Comment (taranveersingh93) [inline src/domUpdates.js:71]: thanks for making this dynamic

File: src/domUpdates.js [hunk 9]
```diff
@@ -181,7 +184,7 @@ const addScrollBar = (element) => {
 const populateInstructions = (recipe) => {
   const instructions = getInstructionHTML(recipe)
   const instructionSection= document.querySelector('#recipeInstructions')
-  instructionSection.innerHTML = `<p>Method</p>
+  instructionSection.innerHTML = `<p>Directions</p>
```
Comment (taranveersingh93) [inline src/domUpdates.js:194]: Thanks for the change

File: src/domUpdates.js [hunk 10]
```diff
@@ -230,18 +233,27 @@ const createIngredientsHTML = ingredients => {
...
-      <input id="${ingredientLabelName}" type="checkbox" name="${ingredientLabelName}" />
+    <label class="ingredient-label" for="${ingredientLabelName}">
+      <input class="ingredient-input" id="${ingredientLabelName}" type="checkbox" name="${ingredientLabelName}" />
       ${ingredient.amount} ${ingredient.unit} ${ingredient.name}
     </label>
     `;
   });
 }
+
+const searchForRecipes = () => {
```
Comment (taranveersingh93) [inline src/domUpdates.js:258]: LOVE this function.

File: src/recipes.js [hunk 11]
```diff
@@ -35,24 +35,33 @@ const calculateRecipeCost = (recipe, ingredients) => {
 }
 
 const filterRecipesByName = (recipes, name) => {
-  return recipes.filter(recipe => recipe.name.includes(name));
+  return recipes.filter(recipe => recipe.name.toLowerCase().includes(name.toLowerCase()));
```
Comment (taranveersingh93) [inline src/recipes.js:38]: Thank you!

File: src/recipes.js [hunk 12]
```diff
@@ -35,24 +35,33 @@ const calculateRecipeCost = (recipe, ingredients) => {
...
 const filterRecipesByIngredient = (recipes, searchedIngredient, ingredientData) => {
-  const matchingIngredients = ingredientData.filter(ingredient => ingredient.name.includes(searchedIngredient));
+  const matchingIngredients = ingredientData.filter(ingredient => ingredient.name?.toLowerCase().includes(searchedIngredient.toLowerCase()));
   const matchingIngredientsID = matchingIngredients.map(ingredient => ingredient.id);
   const matchedRecipes = recipes.filter(recipe => recipe.ingredients.some(ingredient => matchingIngredientsID.includes(ingredient.id)));
   return matchedRecipes;
 };
 
+const searchRecipes = (allRecipes, allIngredients, userSearch) => {
+  const searchedItems = userSearch.split(' ')
```
Comment (taranveersingh93) [inline src/recipes.js]: This step was genius

File: src/recipes.js [hunk 13]
```diff
@@ -35,24 +35,33 @@ const calculateRecipeCost = (recipe, ingredients) => {
...
+  const recipesByName = searchedItems.flatMap(searchWord => filterRecipesByName(allRecipes, searchWord))
+  const recipesByIngredient = searchedItems.flatMap(searchWord => filterRecipesByIngredient(allRecipes, searchWord, allIngredients))
+  const foundRecipes = [...recipesByName, ...recipesByIngredient]
```
Comment (taranveersingh93) [inline src/recipes.js]: NICE!

**Review by taranveersingh93 — CHANGES_REQUESTED** [review]

File: src/styles.css (hunk 1 above)
Comment (lauraguerra1) [reply src/styles.css]: removed these comments

File: src/recipes.js (hunk 2 above)
Comment (lauraguerra1) [reply src/recipes.js]: Thank you !!

File: src/domUpdates.js (hunk 3 above)
Comment (lauraguerra1) [reply src/domUpdates.js]: Resolved in merge conflict - these are gone now!

File: src/scripts.js (hunk 4 above)
Comment (lauraguerra1) [reply src/scripts.js]: resolved in merge conflict! These are gone now

File: test/recipe-test.js (hunk 5 above)
Comment (lauraguerra1) [reply test/recipe-test.js:311]: Thank you!

File: test/recipe-test.js (hunk 6 above)
Comment (lauraguerra1) [reply test/recipe-test.js:294]: Great points, added a conditional and tests for this -> 

```
 it('should return an empty array for non matches', () => {
    let filteredRecipes = searchRecipes(sampleRecipeData, sampleIngredientsData, 'no match string');
    let otherFilteredRecipes = searchRecipes(sampleRecipeData, sampleIngredientsData, '')
    assert.deepEqual(filteredRecipes, []);
    assert.deepEqual(otherFilteredRecipes, [])
  });

  it('should return all recipes if only a space is entered', () => {
    let allRecipes = searchRecipes(sampleRecipeData, sampleIngredientsData, ' ');
    assert.deepEqual(allRecipes, sampleRecipeData)
  }) 
```

Made a new card for resetting the search

File: dist/index.html (hunk 7 above)
Comment (lauraguerra1) [reply dist/index.html:11]: thank you!

File: src/domUpdates.js (hunk 8 above)
Comment (lauraguerra1) [reply src/domUpdates.js:71]: Yes! Will be helpful in future iterations, so I wanted to reuse your logic!

**Review by mbenfowler — APPROVED** [review]

**Comment by mbenfowler** [comment]
We can put this in a follow-up, but I did notice that if you search for a word that then has a space after it, the whole list of recipes are shown.

**Review by taranveersingh93 — APPROVED** [review]
Great work on this one with all the moving parts. Comprehensive changes
